File: src/clj_see/expression.clj
(ns clj-see.expression
  (:require [clj-see.util :as util]))

(defn crossover [expression-1 path-1 expression-2 path-2]
  (let [snippet-1 (util/list-get-in expression-1 path-1)
        snippet-2 (util/list-get-in expression-2 path-2)
        new-expression-1 (util/list-assoc-in expression-1
                                             path-1
                                             snippet-2)
        new-expression-2 (util/list-assoc-in expression-2
                                             path-2
                                             snippet-1)]
    [new-expression-1 new-expression-2]))

(defn mutate [expression path f]
  (let [snippet (util/list-get-in expression path)
        mutated-snippet (f snippet)]
    (util/list-assoc-in expression path mutated-snippet)))

File: src/expression.py
from src.util import *

def crossover(expression1, path1, expression2, path2):
    snippet1 = get_in(expression1, path1)
    snippet2 = get_in(expression2, path2)
    new_expression1 = assoc_in(expression1, path1, snippet2)
    new_expression2 = assoc_in(expression2, path2, snippet1)

    return new_expression1, new_expression2

def mutate(expression, path, mutate):
    snippet = get_in(expression, path)
    mutated_snippet = mutate(snippet)
    new_expression = assoc_in(expression, path, mutated_snippet)

    return new_expression
